File: edoor/edoor/report/general_journal_transaction/general_journal_transaction.js
// Copyright (c) 2023, Tes Pheakdey and contributors
// For license information, please see license.txt
 
 
frappe.query_reports["General Journal Transaction"] = {
	"filters": [
		{
			fieldname: "property",
			label: "Property",
			fieldtype: "Link",
			options:"Business Branch",
			default:frappe.defaults.get_user_default("business_branch") ,
			"reqd": 1,
			"on_change": function (query_report) {
				const property = frappe.query_report.get_filter_value("property")
				const business_source_filter =frappe.query_report.get_filter('business_source');
				business_source_filter.df.get_query = function() {
					return {
						filters: {
							"property": property
						}
					};
				};

				//room type
				const room_type_filter =frappe.query_report.get_filter('room_type');
				room_type_filter.df.get_query = function() {
					return {
						filters: {
							"property": property
						}
					};
				};

				//set fitler city ledger
				const city_ledger =frappe.query_report.get_filter('city_ledger');
				city_ledger.df.get_query = function() {
					return {
						filters: {
							"property": property
						}
					};
				};
				//set filter reservation
				frappe.query_report.get_filter('reservation').df.get_query = function() {
					return {
						filters: {
							"property": property
						}
					};
				};
				
				//set filter for stay
				frappe.query_report.get_filter('reservation_stay').df.get_query = function() {
					return {
						filters: {
							"property": property
						}
					};
				};
				
			 

 

				 
			},
		},
		{
			"fieldname":"start_date",
			"label": __("Start Date"),
			"fieldtype": "Date",
			default: new Date( (new Date()).getFullYear(), ( new Date()).getMonth(), 1),
			"reqd": 1,
			"on_change": function (query_report) {},
		},
		{
			"fieldname":"end_date",
			"label": __("End Date"),
			"fieldtype": "Date",
			default: new Date((new Date()).getFullYear(), (new Date()).getMonth() + 1, 0),
			"on_change": function (query_report) {},
			"reqd": 1
		},

		{
			"fieldname": "room_type",
			"label": __("Room Type"),
			"fieldtype": "Link",
			"options":"Room Type",
			"on_change": function (query_report) {},
			
		},
		{
			"fieldname": "business_source_type",
			"label": __("Business Source Type"),
			"fieldtype": "Link",
			"options":"Business Source Type",
			"on_change": function (query_report) {},
		},
		{
			"fieldname": "business_source",
			"label": __("Business Source"),
			"fieldtype": "Link",
			"options":"Business Source",
			"on_change": function (query_report) {},
		},
		{
			"fieldname": "guest_type",
			"label": __("Guest Type"),
			"fieldtype": "Link",
			"options":"Customer Group",
			"on_change": function (query_report) {},
		},
		{
			"fieldname": "guest",
			"label": __("Guest"),
			"fieldtype": "Link",
			"options":"Customer",
			"on_change": function (query_report) {},
		},
		{
			"fieldname": "city_ledger",
			"label": __("City Ledger"),
			"fieldtype": "Link",
			"options":"City Ledger",
			"on_change": function (query_report) {},
		},
		{
			"fieldname": "reservation",
			"label": __("Reservation"),
			"fieldtype": "Link",
			"options":"Reservation",
			"on_change": function (query_report) {},
		},
		{
			"fieldname": "reservation_stay",
			"label": __("Reservation Stay"),
			"fieldtype": "Link",
			"options":"Reservation Stay",
			"on_change": function (query_report) {},
		},
		{
			"fieldname": "account_code",
			"label": __("Account Code"),
			"fieldtype": "Link",
			"options":"Account Code",
			"on_change": function (query_report) {},
		},
		{
			"fieldname": "account_category",
			"label": __("Account Category"),
			"fieldtype": "MultiSelectList",
			 
			get_data: function(txt) {
				return frappe.db.get_link_options('Account Category', txt);
			},
			"on_change": function (query_report) {},
		},
 
		{
			"fieldname": "ledger_type",
			"label": __("Ledger Type"),
			"fieldtype": "MultiSelectList",
			"options":[
				{"description":"Guest Ledger", "value":"Reservation Folio"},
				{"description":"Deposit Ledger", "value":"Deposit Ledger"},
				{"description":"Desk Folio", "value":"Desk Folio"},
				{"description":"City Ledger", "value":"City Ledger"},
				{"description":"Payable Ledger", "value":"Payable Ledger"},
				{"description":"F&B", "value":"Cashier Shift"}
			],
			"on_change": function (query_report) {},
			hide_in_filter:1,
		},
		 
		{
			"fieldname": "show_summary",
			"label": __("Show Summary"),
			"fieldtype": "Check",
			"default":1,
			hide_in_filter:1,
			"on_change": function (query_report) {},
		},
		{
			"fieldname": "show_package_breakdown",
			"label": __("Show Package Breakdown"),
			"fieldtype": "Check",
			"default":1,
			hide_in_filter:1,
			"on_change": function (query_report) {},
		},
		{
			"fieldname": "show_all_breakdown",
			"label": __("Show All Breakdown"),
			"fieldtype": "Check",
			"default":0,
			hide_in_filter:1,
			"on_change": function (query_report) {},
		},
		 
		 
	],
	onload: function(report) {
		report.page.add_inner_button ("Preview Report", function () {
			frappe.query_report.refresh();
		});
		 
		let currentUrl = window.location.href;

	// Create a URLSearchParams object
	let urlParams = new URLSearchParams(window.location.search);

	// Get the value of the 'auto_refresh' parameter
	let autoRefreshValue = urlParams.get('auto_refresh');
	if (autoRefreshValue==1){
		setTimeout(function(){
			frappe.query_report.refresh();
		},300)
		
		
	}
	},
	"formatter": function(value, row, column, data, default_formatter) {
		const origninal_value = value  || 0
		value = default_formatter(value, row, column, data);
	 
 
		value = value.toString().replace("style='text-align: right'","style='text-align: " + column.align + "'");	
	 
 
		if (
			(column.fieldtype || "") == "Int" || 
			((column.fieldtype || "") == "Percent") ||
			((column.fieldtype || "") == "Currency" ) 
		) 
		{
			if(origninal_value==0){
				return "-"
			}
		} 


		
		if ((data && data.is_group==1) || (data && data.is_total_row==1)) {
			
			value = $(`<span>${value}</span>`);

			var $value = $(value).css("font-weight", "bold");
			

			value = $value.wrap("<p></p>").parent().html();
		}
		
		if(column.fieldtype=="Link"){
			var parser = new DOMParser(); // create a DOMParser object
			var doc = parser.parseFromString(value, "text/html"); // parse the string into a document object
			var element = doc.querySelector("a");
			if (element){
				value = "<a  class='link' data-name='" + element.dataset.name + "' data-doctype='" + element.dataset.doctype+  "'>" + element.textContent + "</a>";
			}
			
		}

	 
 
		return value;
	},
	
};
 

File: edoor/edoor/report/general_journal_transaction/general_journal_transaction.py
# Copyright (c) 2023, Tes Pheakdey and contributors
# For license information, please see license.txt

import frappe
 

def execute(filters=None):
 
	
	 
	 


	data = get_report_data(filters)
	summary = None
	if filters.show_summary ==1:
		summary = get_report_summary(data)

	return get_columns(filters), data, None,None, summary


def get_columns(filters):
	return[
		{"fieldname":"ledger","label":"Ledger/Date", "width":250},
		{"fieldname":"name", "label":"Tran #","fieldtype":"Link","options":"Folio Transaction", "width":150,},
		{"fieldname":"account", "label":"Account", "width":400},
		{"fieldname":"debit", "label":"Debit","fieldtype":"Currency", "width":100,"align":"right"},
		{"fieldname":"credit", "label":"Credit","fieldtype":"Currency", "width":100,"align":"right"},
		{"fieldname":"net_total", "label":"Net Total","fieldtype":"Currency", "width":100,"align":"right"},
	]

def get_data(filters):
	sql="""
		select 
			name,
			transaction_type, 
			date_format(posting_date,'%%d-%%m-%%Y') as ledger, 
			concat(account_code,'-',report_description ) as account, 
			if(type='Debit',amount,0) as debit,
			if(type='Debit',0,amount) as credit,
			1 as indent
		from `tabFolio Transaction`
		where
			property = %(property)s and 
			posting_date between %(start_date)s and %(end_date)s 
		"""
	# add additional filter
	if filters.ledger_type:
		sql = sql + " and transaction_type in %(ledger_type)s "
	if filters.room_type:
		sql = sql + " and room_type_id = %(room_type)s "

	if filters.business_source:
		sql = sql + " and business_source = %(business_source)s "
	
	if filters.guest:
		sql = sql + " and guest = %(guest)s "

	if filters.guest_type:
		sql = sql + " and guest_type = %(guest_type)s "
	
	if filters.business_source_type:
		sql = sql + " and business_source_type = %(business_source_type)s "
		
	if filters.reservation:
		sql = sql + " and reservation = %(reservation)s "

	if filters.reservation_stay:
		sql = sql + " and reservation_stay = %(reservation_stay)s "


	if filters.account_category:

		sql = sql + " and account_category = %(account_category)s "

	if filters.city_ledger:
		sql = sql + " and (transaction_number = %(city_ledger)s or (target_transaction_type='City Ledger' and target_transaction_number=%(city_ledger)s))"

	if filters.account_code:
		if filters.account_code != 'All Account Code':
			filters.account_codes =   get_account_codes(filters.account_code)
			sql = sql + " and account_code in %(account_codes)s"
 
	sql = sql + """
	order by 
		posting_date,
		name
	"""
	
	return  frappe.db.sql(sql, filters ,as_dict =1)

def get_report_data(filters):
	data = get_data(filters)
	report_data = []
	for l in ledger_type():
		transaction_data = [d for d in data if d["transaction_type"] == l["value"]]
		if len(transaction_data)> 0:
			report_data.append({
				"ledger":l["label"],
				"indent":0,
				"is_group":1,
				"debit": sum([d["debit"] for d in transaction_data]),
				"credit": sum([d["credit"] for d in transaction_data]),
				"net_total": sum([d["debit"] - d["credit"] for d in transaction_data])
			})
			report_data = report_data + transaction_data

			#add total row
			report_data.append({
					"ledger":"Total",
					"indent":1,
					"is_group":1,
					"debit": sum([d["debit"] for d in transaction_data]),
					"credit": sum([d["credit"] for d in transaction_data]),
					"net_total": sum([d["debit"] - d["credit"] for d in transaction_data])

			})


	if len(data)> 0:
		report_data.append({})
		report_data.append({
					"ledger":"Grand Total",
					"indent":0,
					"is_group":1,
					"debit": sum([d["debit"] for d in data]),
					"credit": sum([d["credit"] for d in data]),
					"net_total": sum([d["debit"] - d["credit"] for d in data])

			})



	return report_data

def ledger_type():
	return [
		{"label":"Guest Ledger", "value":"Reservation Folio"},
		{"label":"Deposit Ledger", "value":"Deposit Ledger"},
		{"label":"Desk Folio", "value":"Desk Folio"},
		{"label":"City Ledger", "value":"City Ledger"},
		{"label":"Payable Ledger", "value":"Payable Ledger"},
		{"label":"F&B", "value":"Cashier Shift"}
	]

def get_report_summary(data):
	return [
		{"label":"Debit",  "datatype": "Currency", "value":sum([d["debit"] for d in data if "debit" in d  and not "is_group" in d]), "indicator":"blue"},
		{"label":"Credit", "datatype": "Currency", "value":sum([d["credit"] for d in data  if "credit" in d and not "is_group" in d]), "indicator":"red"},
		{"label":"Net Total",  "datatype": "Currency","value":sum([d["debit"] - d["credit"] for d in data if  "debit" in d  and "credit" in d and not "is_group" in d]), "indicator":"green"}
	]

def get_account_codes(account_code):
	account_codes = [account_code]
	d = frappe.get_doc("Account Code", account_code)
	if d.is_group:
		account_codes = account_codes + get_account_code_children(account_code)
	return account_codes
		
def get_account_code_children(parent_account_code):
	
	account_codes = []
	sql = "select name,is_group from `tabAccount Code` where parent_account_code = '{}'".format(parent_account_code)
	data = frappe.db.sql(sql,as_dict=1)
	 
	for d in data:
		account_codes.append(d["name"])
		if d["is_group"]==1:
			
			x =  get_account_code_children(d["name"])
			
			account_codes = account_codes + x
	
	return account_codes
